File: src/main/java/io/scriptor/conf/Parser.java
package io.scriptor.conf;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.NoSuchElementException;

public final class Parser {

    private final InputStream stream;
    private int buffer;
    private Token token;

    public Parser(final @NotNull InputStream stream) throws IOException {
        this.stream = stream;
        this.buffer = stream.read();
        this.token = next();
    }

    public @NotNull Node<?> parse() throws IOException {
        if (at(TokenType.OTHER, "{"))
            return parseObject();
        if (at(TokenType.OTHER, "["))
            return parseArray();

        if (at(TokenType.STRING, TokenType.SYMBOL))
            return new StringNode(skip().string());
        if (at(TokenType.INTEGER))
            return new IntegerNode(skip().integer());
        if (at(TokenType.FLOATING))
            return new FloatingNode(skip().floating());

        if (at("true", "false"))
            return new BooleanNode(Boolean.parseBoolean(skip().string()));

        throw new NoSuchElementException();
    }

    public @NotNull ObjectNode parseObject() throws IOException {
        final var node = new ObjectNode();

        expect(TokenType.OTHER, "{");
        while (!at(TokenType.OTHER, "}")) {
            final var key = expect(TokenType.SYMBOL, TokenType.STRING).string();
            expect(TokenType.OTHER, ":");
            final var value = parse();

            node.put(key, value);

            if (!at(TokenType.OTHER, "}"))
                expect(TokenType.OTHER, ",");
        }
        expect(TokenType.OTHER, "}");

        return node;
    }

    public @NotNull ArrayNode parseArray() throws IOException {
        final var node = new ArrayNode();

        expect(TokenType.OTHER, "[");
        while (!at(TokenType.OTHER, "]")) {
            final var value = parse();

            node.add(value);

            if (!at(TokenType.OTHER, "]"))
                expect(TokenType.OTHER, ",");
        }
        expect(TokenType.OTHER, "]");

        return node;
    }

    private boolean at(final @NotNull TokenType type, final @NotNull String value) {
        return token.type() == type && token.string().equals(value);
    }

    private boolean at(final @NotNull TokenType... types) {
        for (final var type : types)
            if (token.type() == type)
                return true;
        return false;
    }

    private boolean at(final @NotNull String... values) {
        for (final var value : values)
            if (token.string().equals(value))
                return true;
        return false;
    }

    private void expect(final @NotNull TokenType type, final @NotNull String value) throws IOException {
        if (token.type() == type && token.string().equals(value)) {
            token = next();
            return;
        }
        throw new NoSuchElementException("type=%s, value=%s".formatted(type, value));
    }

    private @NotNull Token expect(final @NotNull TokenType... types) throws IOException {
        for (final var type : types)
            if (token.type() == type)
                return skip();
        throw new NoSuchElementException("types=%s".formatted(Arrays.toString(types)));
    }

    private @NotNull Token expect(final @NotNull String... values) throws IOException {
        for (final var value : values)
            if (token.string().equals(value))
                return skip();
        throw new NoSuchElementException("values=%s".formatted(Arrays.toString(values)));
    }

    private @NotNull Token skip() throws IOException {
        final var x = token;
        token = next();
        return x;
    }

    private void read() throws IOException {
        buffer = stream.read();
    }

    private @NotNull Token next() throws IOException {
        var state = State.NONE;

        final var value = new StringBuilder();

        var radix    = 10;
        var floating = false;

        while (buffer >= 0) {
            switch (state) {
                case NONE -> {
                    switch (buffer) {
                        case '#' -> {
                            state = State.COMMENT;
                            read();
                        }
                        case '0' -> {
                            read();
                            switch (buffer) {
                                case 'b' -> {
                                    state = State.NUMBER;
                                    radix = 0b10;
                                    read();
                                }
                                case 'x' -> {
                                    state = State.NUMBER;
                                    radix = 0x10;
                                    read();
                                }
                                default -> {
                                    state = State.NUMBER;
                                    radix = 010;
                                    value.appendCodePoint('0');
                                }
                            }
                        }
                        case '"' -> {
                            state = State.STRING;
                            read();
                        }
                        default -> {
                            if (Character.isWhitespace(buffer)) {
                                read();
                                break;
                            }
                            if (Character.isDigit(buffer)) {
                                state = State.NUMBER;
                                break;
                            }
                            if (Character.isLetter(buffer)) {
                                state = State.SYMBOL;
                                break;
                            }
                            value.appendCodePoint(buffer);
                            read();
                            return new Token(TokenType.OTHER, value.toString(), 0L, 0.0D);
                        }
                    }
                }
                case COMMENT -> {
                    if (buffer == '\n')
                        state = State.NONE;
                    read();
                }
                case SYMBOL -> {
                    if (Character.isLetterOrDigit(buffer)) {
                        value.appendCodePoint(buffer);
                        read();
                        break;
                    }
                    return new Token(TokenType.SYMBOL, value.toString(), 0L, 0.0D);
                }
                case NUMBER -> {
                    if (radix == 10 && !floating && buffer == '.') {
                        floating = true;
                        value.appendCodePoint(buffer);
                        read();
                        break;
                    }
                    if (!floating && buffer == 'u') {
                        read();
                        final var string = value.toString();
                        return new Token(TokenType.INTEGER, string, Long.parseUnsignedLong(string, radix), 0.0D);
                    }
                    if (Character.isLetterOrDigit(buffer)) {
                        value.appendCodePoint(buffer);
                        read();
                        break;
                    }
                    final var string = value.toString();
                    if (floating)
                        return new Token(TokenType.FLOATING, string, 0L, Double.parseDouble(string));
                    return new Token(TokenType.INTEGER, string, Long.parseLong(string, radix), 0.0D);
                }
                case STRING -> {
                    if (buffer == '"') {
                        read();
                        return new Token(TokenType.STRING, value.toString(), 0L, 0.0D);
                    }
                    value.appendCodePoint(buffer);
                    read();
                }
            }
        }

        return new Token(TokenType.EOF, "", 0L, 0.0D);
    }
}

File: src/main/java/io/scriptor/conf/Parser.kt
package io.scriptor.conf

import io.scriptor.util.Log.format
import java.io.InputStream

class Parser {

    private val stream: InputStream

    private var buffer: Int
    private var token: Token

    constructor(stream: InputStream) {
        this.stream = stream
        this.buffer = stream.read()
        this.token = next()
    }

    fun parse(): Node<*> = when {
        at(TokenType.OTHER, "{") -> parseObject()
        at(TokenType.OTHER, "[") -> parseArray()
        at(TokenType.STRING, TokenType.SYMBOL) -> StringNode(skip().string)
        at(TokenType.INTEGER) -> IntegerNode(skip().integer)
        at(TokenType.FLOATING) -> FloatingNode(skip().floating)
        at("true", "false") -> BooleanNode(skip().string.toBoolean())
        else -> throw NoSuchElementException()
    }

    fun parseObject(): ObjectNode {
        val node = ObjectNode()

        expect(TokenType.OTHER, "{")
        while (!at(TokenType.OTHER, "}")) {
            val key = expect(TokenType.SYMBOL, TokenType.STRING).string
            expect(TokenType.OTHER, ":")
            node[key] = parse()

            if (!at(TokenType.OTHER, "}")) {
                expect(TokenType.OTHER, ",")
            }
        }
        expect(TokenType.OTHER, "}")

        return node
    }

    fun parseArray(): ArrayNode {
        val node = ArrayNode()

        expect(TokenType.OTHER, "[")
        while (!at(TokenType.OTHER, "]")) {
            node.add(parse())

            if (!at(TokenType.OTHER, "]")) {
                expect(TokenType.OTHER, ",")
            }
        }
        expect(TokenType.OTHER, "]")

        return node
    }

    private fun at(type: TokenType, value: String): Boolean {
        return token.type == type && token.string == value
    }

    private fun at(vararg types: TokenType): Boolean {
        for (type in types) {
            if (token.type == type) {
                return true
            }
        }
        return false
    }

    private fun at(vararg values: String): Boolean {
        for (value in values) {
            if (token.string == value) {
                return true
            }
        }
        return false
    }

    private fun expect(type: TokenType, value: String) {
        if (token.type == type && token.string == value) {
            token = next()
            return
        }
        throw NoSuchElementException(format("type=%s, value=%s", type, value))
    }

    private fun expect(vararg types: TokenType): Token {
        for (type in types) {
            if (token.type == type) {
                return skip()
            }
        }
        throw NoSuchElementException(format("types=%s", types.contentToString()))
    }

    private fun expect(vararg values: String): Token {
        for (value in values) {
            if (token.string == value) {
                return skip()
            }
        }
        throw NoSuchElementException(format("values=%s", values.contentToString()))
    }

    private fun skip(): Token {
        val x = token
        token = next()
        return x
    }

    private fun read() {
        buffer = stream.read()
    }

    private fun next(): Token {
        var state = State.NONE

        val value = StringBuilder()

        var radix = 10
        var floating = false

        while (buffer >= 0) when (state) {
            State.NONE -> when (buffer) {
                '#'.code -> {
                    state = State.COMMENT
                    read()
                }

                '0'.code -> {
                    state = State.NUMBER
                    read()

                    when (buffer) {
                        'b'.code -> {
                            radix = 2
                            read()
                        }

                        'x'.code -> {
                            radix = 16
                            read()
                        }

                        else -> {
                            radix = 8
                            value.appendCodePoint('0'.code)
                        }
                    }
                }

                '"'.code -> {
                    state = State.STRING
                    read()
                }

                else -> when {
                    buffer.toChar().isWhitespace() -> {
                        read()
                    }

                    buffer.toChar().isDigit() -> {
                        state = State.NUMBER
                    }

                    buffer.toChar().isLetter() -> {
                        state = State.SYMBOL
                    }

                    else -> {
                        value.appendCodePoint(buffer)
                        read()
                        return Token(TokenType.OTHER, value.toString(), 0U, 0.0)
                    }
                }
            }

            State.COMMENT -> {
                if (buffer == '\n'.code) {
                    state = State.NONE
                }
                read()
            }

            State.SYMBOL -> when {
                buffer.toChar().isLetterOrDigit() -> {
                    value.appendCodePoint(buffer)
                    read()
                }

                else -> return Token(TokenType.SYMBOL, value.toString(), 0U, 0.0)
            }

            State.NUMBER -> when {
                radix == 10 && !floating && buffer == '.'.code -> {
                    value.appendCodePoint(buffer)
                    read()

                    floating = true
                }

                !floating && buffer == 'u'.code -> {
                    read()

                    val string = value.toString()
                    return Token(TokenType.INTEGER, string, string.toULong(radix), 0.0)
                }

                buffer.toChar().isLetterOrDigit() -> {
                    value.appendCodePoint(buffer)
                    read()
                }

                else -> {
                    val string = value.toString()

                    return if (floating) {
                        Token(TokenType.FLOATING, string, 0U, string.toDouble())
                    } else {
                        Token(TokenType.INTEGER, string, string.toLong(radix).toULong(), 0.0)
                    }
                }
            }

            State.STRING -> when (buffer) {
                '"'.code -> {
                    read()
                    return Token(TokenType.STRING, value.toString(), 0U, 0.0)
                }

                else -> {
                    value.appendCodePoint(buffer)
                    read()
                }
            }
        }

        return Token(TokenType.EOF, "", 0U, 0.0)
    }
}
